### PR TITLE
Add message reply based on contacts

### PR DESCRIPTION
**Pull request by hegocre**
Implemented the ability to choose which contacts to reply.

This needs testing, but looks like it's working for me.

It's only an initial approach, interface could be changed to something nicer and some code might need to be moved to more appropriate places.

### CODE REVIEW

**Comment by adeekshith** [comment]
Looks great! Thank you @hegocre 😃
Yet to check the code but it is working for me.

Screenshots:

|     |     |
| ------------- |:-------------:|
| Screenshot_1623123784     | Screenshot_1623123788 |
| Screenshot_1623123792     | Screenshot_1623123838     |

**Comment by hegocre** [comment]
> Looks great!

That's nice! Let me know if there's something that needs to be changed!

File: app/src/main/java/com/parishod/watomatic/NotificationService.java [hunk 1]
```diff
@@ -40,6 +39,32 @@ private boolean canReply(StatusBarNotification sbn){
                 canSendReplyNow(sbn);
     }
 
+    private boolean shouldReply(StatusBarNotification sbn){
+        PreferencesManager prefs = PreferencesManager.getPreferencesInstance(this);
+        boolean isGroup = sbn.getNotification().extras.getBoolean("android.isGroupConversation");
```
Comment (adeekshith) [inline app/src/main/java/com/parishod/watomatic/NotificationService.java:44]: Everything is good except group detection. This should work ideally but there is an issue where images messages in group are not being detected as `isGroupConversation` sometimes (#181).

So there is an extra logic to detect that: [src](https://github.com/adeekshith/watomatic/blob/eac18a2877718eb3a2f07493f91b8d6dab250900/app/src/main/java/com/parishod/watomatic/NotificationService.java#L149-L156)

May be we should move that logic to a separate method and reuse it.
Or I will merge this PR as is this evening and raise another one fixing this.

File: app/src/main/java/com/parishod/watomatic/model/CustomRepliesData.java [hunk 2]
```diff
@@ -116,13 +116,11 @@ public String getOrElse(String defaultText) {
     }
 
     public String getTextToSendOrElse (String defaultTextToSend) {
-        String currentText = get();
+        String currentText = getOrElse(thisAppContext.getString(R.string.auto_reply_default_message));
         if (preferencesManager.isAppendWatomaticAttributionEnabled()) {
             currentText += "\n\n"+ RTL_ALIGN_INVISIBLE_CHAR + thisAppContext.getString(R.string.sent_using_Watomatic);
         }
-        return (currentText != null)
```
Comment (adeekshith) [inline app/src/main/java/com/parishod/watomatic/model/CustomRepliesData.java:123]: 👍

File: app/src/main/java/com/parishod/watomatic/model/CustomRepliesData.java [hunk 3]
```diff
@@ -116,13 +116,11 @@ public String getOrElse(String defaultText) {
...
-                ? currentText
-                : defaultTextToSend;
+        return currentText;
```
Comment (adeekshith) [inline app/src/main/java/com/parishod/watomatic/model/CustomRepliesData.java:123]: 👍

File: app/src/main/java/com/parishod/watomatic/NotificationService.java (hunk 1 above)
Comment (hegocre) [reply app/src/main/java/com/parishod/watomatic/NotificationService.java:44]: I tried using that logic, but didn't manage to do it right, so went the easy way as a temporary fix.

Comment (hegocre) [reply app/src/main/java/com/parishod/watomatic/NotificationService.java:44]: I've read the issue you mentioned, if we are gonna add a new option to do extra checks in settings, i think this should be changed on the new PR

Comment (adeekshith) [reply app/src/main/java/com/parishod/watomatic/NotificationService.java:44]: No probs 😊
I'll merge this PR in the evening ([CST](https://www.wolframalpha.com/input/?i=Current+time+in+lllinois%2C+USA)) as is and make that fix before I forget. Great work!! 🙌

Also, as you are already good with Kotlin, feel free to use that in the future (anything is fine though). I am planning to move this project to Kotlin in the near future anyways as it is getting difficult to stick to Java in the Android world.

Comment (adeekshith) [reply app/src/main/java/com/parishod/watomatic/NotificationService.java:44]: > I've read the issue you mentioned, if we are gonna add a new option to do extra checks in settings, i think this should be changed on the new PR

Yep, Had to do a hotfix for a previous release. As soon as that is done, will make a PR to fix this group issue.

Comment (hegocre) [reply app/src/main/java/com/parishod/watomatic/NotificationService.java:44]: > No probs blush
I'll merge this PR in the evening (CST) as is and make that fix before I forget. Great work!! raised_hands

Sounds nice

> Also, as you are already good with Kotlin, feel free to use that in the future (anything is fine though). I am planning to move this project to Kotlin in the near future anyways as it is getting difficult to stick to Java in the Android world.

Yay!

**Review by adeekshith — APPROVED** [review]

**Comment by fikimaul** [comment]
Nice work bro @adeekshith @hegocre 🚀🚀

**Comment by tiptoptom** [comment]
I think this is like a whitelist. What about adding also a blacklist feature?

**Comment by hegocre** [comment]
Yes, it has both, as @adeekshith commented

**Comment by OLLI-S** [comment]
When will that be available?
I have **1.19.1** and here the feature is not yet available.

**Comment by adeekshith** [comment]
> When will that be available?
> I have **1.19.1** and here the feature is not yet available.

Hi @OLLI-S 
This should be available in a week. Already tagged and tested. Uploading to Google play today. Might take 2 to 4 days to get it approved and then a few more days to roll out to everyone.

**Comment by OLLI-S** [comment]
@adeekshith Thank you for the info

**Comment by adeekshith** [comment]
Uploaded to Google Play. It warns that users will have to manually update this as this requests a new Contacts permission. Just concerned about it a little I hope it should be okay 😕

`Users that have the APK with version code 24 may need to accept the android.permission.READ_CONTACTS permission, which may result in them not upgrading to this version of the app.`

**Comment by OLLI-S** [comment]
Update received some days ago...

**Pull request merged**